src/js/angular/guides/tour-lib-services/shepherd.service.js

```diff
@@ -0,0 +1,543 @@
+        }
+        if (stepsDescriptions.length > 2) {
+            for (let index = 1; index < stepsDescriptions.length - 1; index++) {
+                guide.addStep(this._getMiddleStep(guide, stepsDescriptions, index));
+            }
+        }
+        if (stepsDescriptions.length > 1) {
+            guide.addStep(this._getLastStep(guide, stepsDescriptions));
+        }
+    }
+
+    /**
+     * Starts <code>guide</code> from <code>startStepId</code>
+     * @param guide - the guid to be started.
+     * @param startStepId - star step id. If it is not present the <code>guide</code> will be started from beginning.
+     * @private
+     */
+    this._startGuide = (guide, startStepId) => {
+        const stepIndex = this._getStepIndex(guide, startStepId);
+        const step = guide.steps[stepIndex];
+        const url = step.options.url;
+        if (!!url) {
+            $location.path(url);
+        }
+
+        GuideUtils.waitFor(step.options.attachTo.element, step.options.maxWaitTime)
+            .then(() => guide.show(stepIndex))
+            .catch(() => guide.cancel());
+    }
+
+    this._subscribeToGuideCanceled = () => {
+        if (!this.guideCancelSubscription) {
+            this.guideCancelSubscription = Shepherd.on('cancel', () => {
+                this._clearLocaleStorage();
+                if (this.onCancel) {
+                    this.onCancel();
+                }
+            });
+        }
+    }
+
+    /**
+     * Creates first step from <code>stepsDescriptions</code>.
+     * @param guide - the guide.
+     * @param stepsDescriptions - array with descriptions of steps.
+     * @returns the first step.
+     * @private
+     */
+    this._getFirstStep = function (guide, stepsDescriptions) {
+        const nextStepDescription = stepsDescriptions.length > 1 ? stepsDescriptions[1] : undefined;
+        return this._toGuideStep(guide, undefined, stepsDescriptions[0], nextStepDescription);
+    }
+
+    /**
+     * Creates middle step from <code>stepsDescriptions</code>.
+     * @param guide - the guide.
+     * @param stepsDescriptions - array with descriptions of steps.
+     * @param indexOfProcessedStep - index of step which have to be created.
+     * @returns the created step.
+     * @private
+     */
+    this._getMiddleStep = (guide, stepsDescriptions, indexOfProcessedStep) => {
+        return this._toGuideStep(guide, stepsDescriptions[indexOfProcessedStep - 1], stepsDescriptions[indexOfProcessedStep], stepsDescriptions[indexOfProcessedStep + 1]);
+    }
+
+    /**
+     * Creates last step from <code>stepsDescriptions</code>.
+     * @param guide - the guide.
+     * @param stepsDescriptions - array with descriptions of steps.
+     * @returns the last step.
+     * @private
+     */
+    this._getLastStep = (guide, stepsDescriptions) => {
+        return this._toGuideStep(guide, stepsDescriptions[stepsDescriptions.length - 2], stepsDescriptions[stepsDescriptions.length - 1]);
```

Comment:
You usually check for some values in the stepsDescriptions before accessing them in the functions above. What's different here?

Reply:
I wrote this function only for readability purpose.